### PR TITLE
Cypress refactor

### PR DESCRIPTION
**Pull request by david-roper**
The following merge request includes:

- A E2E test suite for the OpenDataCapture Web Application
- A mock service worker for mock API calls used in the E2E tests
- Adds data-cy labels to UI elements for future testing purposes

Steps to run the test suite are as follows:

cd /OpenDataCapture/apps/web
pnpm test:dev

Non CLI version
pnpm test

### CODE REVIEW

File: apps/web/cypress/e2e/tags.cy.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+    //navigate to view instruments, select the first tag available with dropdown
+    cy.wait(1000);
+    cy.get('button[data-cy="view-instrument"]').first().click({ force: true });
+    cy.get('div[data-cy="tags-btn-dropdown"]').click();
+    cy.get('li[id*="headlessui-listbox-option-:"]').eq(0).click({ force: true });
+
+    //check that a form now contains that selected tag
+    let text1: string;
+    cy.get('span[class="ui-selected:font-medium"]')
```
Comment (joshunrau) [inline apps/web/cypress/e2e/tags.cy.ts:47]: see overall comments

File: apps/web/cypress/e2e/tags.cy.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        text1 = $span.text();
+      });
+    cy.get('h5[data-cy="instrument-card"]')
+      .eq(0)
+      .should(($h5) => {
+        const spanText1 = $h5.text();
+        expect(spanText1).to.contain(text1);
+      });
+
+    //deselect the first tag
```
Comment (joshunrau) [inline apps/web/cypress/e2e/tags.cy.ts:59]: see overall comments

File: apps/web/cypress/e2e/tags.cy.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    cy.wait(200);
+    cy.get('li[id*="headlessui-listbox-option-:"]').eq(0).click({ force: true });
+    cy.wait(200);
+    cy.get('li[id*="headlessui-listbox-option-:"]').eq(1).click({ force: true });
+
+    //check form contains the second tag
```
Comment (joshunrau) [inline apps/web/cypress/e2e/tags.cy.ts:65]: see overall comments

File: apps/web/cypress/e2e/tags.cy.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        text2 = $span.text();
+      });
+    cy.get('h5[data-cy="instrument-card"]')
+      .eq(0)
+      .should(($h5) => {
+        const spanText2 = $h5.text();
+        expect(spanText2).to.contain(text2);
+      });
+
+    //deselect the second tag, select the third tag
```
Comment (joshunrau) [inline apps/web/cypress/e2e/tags.cy.ts:79]: see overall comments

File: apps/web/cypress/e2e/tags.cy.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        text4 = $span.text();
+      });
+    cy.get('h5[data-cy="instrument-card"]')
+      .eq(0)
+      .should(($h5) => {
+        const spanText4 = $h5.text();
+        expect(spanText4).to.contain(text4);
+      });
+
+    //deselect schizo and psychosis from search
```
Comment (joshunrau) [inline apps/web/cypress/e2e/tags.cy.ts:116]: see overall comments

File: apps/web/cypress/support/commands.ts [hunk 6]
```diff
@@ -14,18 +14,28 @@
...
+  //change wait times based off network speed
+  //a slow network will require a long wait time in order to load all the handlers and mock data
+  cy.wait(2000);
   const form = cy.get('form[data-cy="login-form"]');
   form.get('input[name="username"]').type(username);
   form.get('input[name="password"]').type(password);
   form.get('button[aria-label="Submit Button"]').click();
 });
 
+Cypress.Commands.add('dragTo', { prevSubject: 'element' }, (subject, targetSelector) => {
```
Comment (joshunrau) [inline apps/web/cypress/support/commands.ts:27]: This is cool!

File: apps/web/src/features/subjects/components/SubjectLayout.tsx
Comment (joshunrau) [inline apps/web/src/features/subjects/components/SubjectLayout.tsx:1]: Great!

File: apps/web/src/test/server/browser.ts [hunk 7]
```diff
@@ -1,6 +1,19 @@
 import { setupWorker } from 'msw/browser';
 
 import { authHandlers } from './handlers/auth.handlers';
+import { FormHandlers } from './handlers/form.handlers';
+import { ManageInstrumentHandlers } from './handlers/instrument-manager.handlers';
```
Comment (joshunrau) [inline apps/web/src/test/server/browser.ts:5]: Why is this capitalized?

File: apps/web/src/test/server/browser.ts [hunk 8]
```diff
@@ -1,6 +1,19 @@
...
 import { setupHandlers } from './handlers/setup.handlers';
+import { SubjectHandlers } from './handlers/subject.handlers';
```
Comment (joshunrau) [inline apps/web/src/test/server/browser.ts:7]: Why is this capitalized?

File: apps/web/src/test/server/handlers/summary.handlers.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import type { Summary } from '@open-data-capture/common/summary';
+import { HttpResponse, http } from 'msw';
+
+export const SummaryHandlers = [
```
Comment (joshunrau) [inline apps/web/src/test/server/handlers/summary.handlers.ts:4]: Why is this capitalized?

File: apps/web/src/test/server/handlers/visit.handlers.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
+import type { Visit } from '@open-data-capture/common/visit';
+import { HttpResponse, http } from 'msw';
+
+import { testSubject } from '../stubs';
+
+export const VisitHandler = [
```
Comment (joshunrau) [inline apps/web/src/test/server/handlers/visit.handlers.ts:6]: Why is this capitalized?

File: pnpm-lock.yaml
Comment (joshunrau) [inline pnpm-lock.yaml:1]: Please discard these changes. I cannot on GitHub UI.

**Review by joshunrau — CHANGES_REQUESTED** [review]
Why do you force click all buttons? Do not add force to things unless it is necessary (which as far as I understand, should not be the case anywhere here). If it is necessary, please let me know and I will investigate if there is a problem.

Using `cy.wait` is almost always considered an anti-pattern (https://docs.cypress.io/api/commands/wait#Wait-for-an-arbitrary-period-of-milliseconds). Please remove all usages of this, or provide a justification for why it is necessary.

Please avoid getting DOM elements by non-specific attributes (including, but not limited to, class name, type of the element, etc.). It is not obvious what you are getting and this is extremely easy for me, or somebody else, to break. Please prefer getting items by test id, or in the case of forms, name, which will not change between UI updates.

Prefer getting the element by test ID (or equivalent). If you are getting the element by something that could theoretically be duplicated, but in-practice never will (e.g., text content of buttons to select the year), please mention this is a comment. You can make a PR against JSLibs if you need to add test ids to things.

Under no circumstances depend on implementation details of third party libraries for tests. This will almost certainly break from under us at a certain point

File: pnpm-lock.yaml
Comment (joshunrau) [reply pnpm-lock.yaml:1]: I discarded your changes in package.json (you accepted the wrong changes on merge), but this I can't

File: .gitignore [hunk 11]
```diff
@@ -2,10 +2,12 @@
 node_modules
 .pnp
 .pnp.js
+bun.lockb
 
 # testing
 coverage
 apps/web/cypress/screenshots
+data/
```
[the remark this answers is not in the document]
Comment (david-roper) [reply .gitignore]: fixed

File: .gitignore [hunk 12]
```diff
@@ -2,10 +2,12 @@
 node_modules
 .pnp
 .pnp.js
+bun.lockb
```
[the remark this answers is not in the document]
Comment (david-roper) [reply .gitignore]: removed

File: apps/web/cypress/e2e/add-visit.cy.ts [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+const name = makeid(6);
+
+describe('look at questionaire', () => {
+  it('passes', () => {
+    //navigate to add visit page, fill in subject form
+    cy.login(adminUser.username, adminUser.password);
+    cy.wait(2000);
+    cy.get('button[data-cy="add-visit"]').first().click({ force: true });
```
[the remark this answers is not in the document]
Comment (david-roper) [reply apps/web/cypress/e2e/add-visit.cy.ts:8]: certain UI elements require force: true to be clicked when they are overlapping another element. 
This scenario the force true was used if ever the add visit modal overlapped the side bar, but not it is unnecessary and removed.

File: apps/web/cypress/e2e/add-visit.cy.ts [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    cy.get('input[name=firstName]').type(name);
+    cy.get('input[name=lastName]').type('testPatient');
+
+    //activate a fill in DOB date reader
+    cy.get('input[class=field-input]').first().type('test');
```
[the remark this answers is not in the document]
Comment (david-roper) [reply apps/web/cypress/e2e/add-visit.cy.ts:13]: The DOB input does not currently have a name variable. It is a weird case where it should be clicked to select dates from the dropdown but is considered a "type into input" to activate said dropdown by cypress. Adding a name to the element will make the test more readable.

File: apps/web/cypress/e2e/add-visit.cy.ts [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    cy.get('svg[data-testid="arrow-up-icon"]').eq(1).click();
+    cy.get('button').contains('1999').click();
+    cy.wait(500);
```
[the remark this answers is not in the document]
Comment (david-roper) [reply apps/web/cypress/e2e/add-visit.cy.ts]: replaced with waiting for element to be visible

**Review by david-roper — COMMENTED** [review]
All comments have now been addressed and resolved in most recent commits

**Review by joshunrau — APPROVED** [review]
much better, thanks

**Pull request merged**